src/components/Router/index.js

```diff
@@ -110,32 +97,89 @@ class MainRouter extends Component {
     this.setState({mobileConnected: true})
   }
 
-  onStepChange = ({step, flow, mobileInitialStep}) => {
-    this.setState({step, flow, mobileInitialStep})
+
+  onFlowChange = (newFlow, mobileInitialStep) => {
+    if (newFlow === "crossDeviceSteps") this.setState({mobileInitialStep})
   }
 
-  buildComponentsList = () => {
-    const params = {
-      flow: this.state.flow,
-      documentType: this.props.documentType,
-      steps: this.props.options.steps
+  render = (props) =>
+      <HistoryRouter {...props}
+        onFlowChange={this.onFlowChange}
+        roomId={this.state.roomId}
+      />
+}
+
+class HistoryRouter extends Component {
+  constructor(props) {
+    super(props)
+    this.state = {
+      flow: 'captureSteps',
+      step: this.props.step || 0,
     }
-    return componentsList(params)
+    this.unlisten = history.listen(this.onHistoryChange)
+    this.setStepIndex(this.state.step, this.state.flow)
   }
 
-  render = (props) => {
-    const components = this.buildComponentsList()
-    return (
+  onHistoryChange = ({state:historyState}) => {
+    this.props.onStepChange(historyState)
+    this.setState({...historyState})
+  }
+
+  componentWillUnmount () {
+    this.unlisten()
+  }
+
+  changeFlowTo = (newFlow) => {
+    const {flow: previousFlow, step: previousStep} = this.state
+    if (previousFlow === newFlow) return
+    this.props.onFlowChange(newFlow, previousStep, previousFlow)
```

Comment:
Why are you passing `previousFlow` here?

Reply:
not really needed to be honest, but usually event callbacks that have to purpose to remind the client a value has changed usually include the previous value as well. but it's not necessary